probe-rs/src/config/sequences/nxp_armv7m.rs

```diff
@@ -177,167 +115,370 @@ impl ArmDebugSequence for MIMXRT10xx {
+        _: probe_rs_target::CoreType,
+        _: Option<u64>,
+    ) -> Result<(), ArmError> {
+        self.simulate_reset_catch.store(false, Ordering::Relaxed);
         Ok(())
     }
 
     fn reset_system(
         &self,
-        interface: &mut dyn ArmProbe,
-        _: crate::CoreType,
-        _: Option<u64>,
+        probe: &mut dyn ArmProbe,
+        core_type: probe_rs_target::CoreType,
+        debug_base: Option<u64>,
     ) -> Result<(), ArmError> {
-        // It's unpredictable to VECTRESET a core if it's not halted and
-        // in debug state.
-        tracing::debug!("Halting MIMXRT11xx core before VECTRESET");
-        let mut dhcsr = Dhcsr(0);
-        dhcsr.set_c_halt(true);
-        dhcsr.set_c_debugen(true);
-        dhcsr.enable_write();
+        // Cache debug system state that may be lost across the reset.
+        let debug_cache = DebugCache::from_target(probe)?;
 
-        interface.write_word_32(Dhcsr::get_mmio_address(), dhcsr.into())?;
-        std::thread::sleep(Duration::from_millis(100));
+        // Make sure that the CM7's SYSRESETREQ isn't ignored by the system
+        // reset controller.
+        self.clear_src_srmr_mask(probe)?;
 
-        // Initial testing showed that a SYSRESET (the default reset approach)
-        // can result in an unreliable programming sequence, particularly if
-        // the target we're reprogramming is interrupting / excepting.
+        // Affect a SYSRESETREQ throught the CM7 to reset the entire system.
+        //
+        // For more information on the SYSRESETREQ response, consult the system
+        // reset controller (SRC) section of the reference manual. This is a
+        // convenient way to perform a whole-system reset.
+        //
+        // Another approach to perform this reset: iterate through all SRC slice controls,
+        // and manually reset them. That should be close to SYSRESETREQ. However, it seems
+        // that there are no slice controls for CM4MEM (LMEM) and CM7MEM (FlexRAM)
+        // so you might not be able to affect a reset on those two domains.
         //
-        // The debug port setup (above) will trap the core(s) after this VECRESET.
-        // Once that trap happens, we're ready to debug / flash.
-        tracing::debug!("Resetting MIMXRT11xx with VECTRESET");
+        // We're about to lose the debug port! We're ignoring missed or incorrect responses.
         let mut aircr = Aircr(0);
         aircr.vectkey();
-        aircr.set_vectreset(true);
-
-        interface
+        aircr.set_sysresetreq(true);
+        probe
             .write_word_32(Aircr::get_mmio_address(), aircr.into())
             .ok();
-        interface.flush().ok();
+        probe.flush().ok();
 
-        std::thread::sleep(Duration::from_millis(100));
+        // If all goes well, we lost the debug port. Thanks, boot ROM. Let's bring it back.
+        //
+        // The ARM communication interface knows how to re-initialize the debug port.
```

Comment:
Maybe we should add `Session::reset_system` and implement the reattach there, if the sequence returns `ReattachRequired`. We have the precedent in erase_chip, and we also have pains in other MCUs.

Reply:
That's all do-able. If we were to return `ReattachRequired` at this line, we would still need to execute all the code beneath this line to set up the reset context. As the debug sequence, I'll need to track which part of a `reset_system` I'm performing, and I'll trust that `reset_system` will be called again after I returned `ReattachRequired`.

In regards to `Session::reset_system`: what about lowering the behavior into [`Core::reset`](https://docs.rs/probe-rs/0.23.0/probe_rs/struct.Core.html#method.reset) or into the various (ARM-only?) cores? Folks are already expecting `Core::reset` to affect a proper system reset.

Reply:
Migrating to `Core::reset` would assume that every possible target can be reset from any of its cores. I don't believe that to be true, multi-core ESP32s look to require operations on both cores (and they aren't even managed by the same debug module).

Reply:
>> I'll trust that reset_system will be called again after I returned ReattachRequired.
>
> Probably not, although debug_core_start would most likely be called.

I see how `ReattachRequired` could suggest that we've _finished_ in the reset, so we only need to re-attach. But for this chip, I would need to return an error like `ReattachAndFinishTheResetRequired`. The difference is that the reset _hasn't finished_, so we would need to do more than re-attach.

Does this help clarify why I would need `reset_system`, or something equivalent, to be called again? `ReattachAndFinishTheResetRequired` is a mouthful, but maybe a new error would be appropriate.

Reply:
Sorry I edited my comment just recently - we can delay this part of the work. It's clear enough why the target needs to be brought out of reset, I just this we can detect this case in `debug_core_start` and do the work there. Maybe I'm wrong here, but anyway let's come back to this part later.